sigdfl.3
.TH sigdfl 3
.SH NAME
sigdfl \- invoke the default action for a signal
.SH SYNTAX
.B #include <sigdfl.h>

int \fBsigdfl(\fIsig\fB)\fR;

int \fBsigdfl_ttou()\fR;
.br
int \fBsigdfl_tstp()\fR;
.br
int \fBsigdfl_ttin()\fR;
.br
int \fBsigdfl_stop()\fR;

int \fBsigdfl_abrt()\fR;

int \fIsig\fP;
.SH DESCRIPTION
.B sigdfl(\fIsig\fB)
has exactly the same effect as what happens 
if the process isn't blocking
.I sig,
has SIG_DFL as the handler for
.I sig,
and then receives
.I sig.
In the case of stop signals
(TSTP, TTIN, TTOU, STOP),
.B sigdfl
will absorb one or more CONT
signals before returning.
.B sigdfl
is safe against
multithreading.

.B sigdfl
has several interesting uses inside signal handlers.
For instance, the handler for SEGV can clean up the tty
and then call
.B sigdfl(\fRSIGSEGV\fB).
Unlike a simple
.B abort(),
this doesn't hide the reason for the core dump from the user.
Similar comments apply to the stop signals.
Note that processes handling a stop signal should
take steps to ensure that they are back in the foreground
after
.B sigdfl
returns.

.B sigdfl
returns 0 normally, -1 with
errno
set in case of various impossible errors.

.B sigdfl_tstp()
is the same as
.B sigdfl(\fRSIGTSTP\fB).
On systems not supporting
SIGTSTP,
.B sigdfl_tstp
returns -1 with errno ENOTTY.
Similar comments apply to
.B sigdfl_stop(),
.B sigdfl_ttin(),
and
.B sigdfl_ttou().
.B sigdfl_abrt()
is a professional version of
.B abort();
it will dump core with the ABRT signal,
even if an ABRT handler is in place.

SIG_DFL
could, in principle, be implemented as
.B sigdfl.
.SH VERSION
sigdfl 1.0, July 18, 1991.
.SH AUTHOR
Placed into the public domain by Daniel J. Bernstein.
.SH "SEE ALSO"
signal(3),
abort(3)


CHANGES
2/9/92: pty 4.0.
2/9/92: Fixed up documentation, hopefully for the last time.
1/29/92: Changed INSTALL.c to report chown root for INSTALL root. Tnx IS.
1/27/92: Added ABRT->IOT ifndef in sigdfl.c.
1/27/92: Changed sigdfl.c to zero the sigvecs rather than setting sv_flags.
1/16/92: pty 4.0, second gamma.
1/13/92: Does it work on an Iris? Nobody knows. 
1/13/92: Added IRIS_UGH_PTYS to ptyget.c. Tnx IRIS, I think.
1/13/92: Changed fatal to warning for tctpgrp failure in ptyslave.c. Tnx IRIS.
1/13/92: ifndef'ed SYSV for ut_host in ptylogs.c. Tnx IRIS.
1/13/92: Added config/sysv.h to distribution. Tnx IRIS.
1/13/92: Added PTYLIBS, noted its use in README. Tnx IRIS.
1/13/92: Replaced CCOPTMZ with OPTMZ.
1/13/92: Noted in CALLS that TTY_TERMIO seems to work. Tnx IRIS.
1/13/92: Adapted tty_initmodes for TTY_TERMIO. Ugh, ugh. Tnx IRIS.
1/13/92: Adapted tty_mungemodes for TTY_TERMIO. Ugh. Tnx IRIS.
1/13/92: Adapted tty_modifymodes for TTY_TERMIO (poorly). Tnx IRIS.
1/13/92: Adapted tty_getmodes and tty_setmodes for TTY_TERMIO. Tnx IRIS.
1/13/92: In ptytty.c used ifdefs for FIONREAD and TIOCEXCL. Tnx IRIS.
1/13/92: Adapted ptytty.h for TTY_TERMIO. Tnx IRIS.
1/13/92: Added TTY_TERMIO option to config/ttyopts.h. Tnx IRIS.
1/13/92: Gave in and made tty_copymodes do a struct copy.
1/13/92: Fixed typo in SECURITY. Tnx KS.
1/13/92: Fixed typo in pty-basic.1. Tnx KS.
1/12/92: Replaced <sgtty.h> with <sys/ioctl.h> in ptytty.h. Tnx IRIS.
1/12/92: Added chmod 700 for all shell scripts in Makefile.
1/12/92: Removed <sys/timeb.h> from wall.c and write.c. Tnx IRIS.
1/12/92: In separate effort got it working on an IRIS. See above.
12/22/91: pty 4.0, first gamma! (Still have to integrate BSD 4.4 version.)
12/22/91: Changed various things in the documentation.
12/22/91: Fixed exit code reporting for, e.g., DIE_NOPTYS.
12/22/91: Changed POSIX to POSIX_SILLINESS. See config/posix.h.
12/20/91: Added CCOPTMZ in Makefile for radixsort.
12/19/91: Added BLURB.
12/19/91: Added TESTS.
12/19/91: Cleaned up README.
12/19/91: Added WTF, noted SunOS 4.0.3 forcefg bug in README.
12/18/91: Fixed bug in sessmenu's sesslist parsing.
12/17/91: Fixed up documentation a bit; referred to checkptys in README.
12/16/91: Added CPP define to Makefile.
12/15/91: Added -xi.
12/15/91: Started machine list in Makefile.
12/15/91: Added EWOULDBLOCK in ptysecure.c for Domain/OS.
12/14/91: Added target-list target.
12/14/91: Added lint target.
12/13/91: Changed master so that it continues child on any reconnect. Tnx SS.
11/24/91: Changed master to run only as effective uid.
11/18/91: Added TERM, USR1, USR2 ignoring to master. Tnx SS.
11/9/91: Added \fP in syntax sections in pty-opts.1, pty.1. Tnx SS.
11/9/91: Added <sys/types.h> before <utmp.h> in {users,who}.c. Domain/OS idiocy.
11/9/91: Fixed dependencies for CHECKCONF and INSTALL.
11/3/91: Added proper .h dependencies in Makefile (finally).
10/27/91: Added exit(0) at end of sesslist.c. Tnx SS.
10/27/91: Fixed ptysecure.c---it didn't undo chmod for some setups. Tnx SS.
10/27/91: Added make clean. Tnx SS.
10/27/91: Added checkptys.8.
10/27/91: Changed checkptys to be installed non-setuid.
10/26/91: Added pty.1, pty-basic.1, pty-opts.1, sess.1, condom.1.
10/26/91: Tossed old man pages.
10/26/91: Added DIE_EXIST for ptysigler.c.
10/26/91: Added DIE_ELSE for ptysigler.c.
10/26/91: Fixed ptymain to die(SETMODES) in an impossible situation.
10/26/91: Changed a few die() values in ptymain.c.
10/25/91: Need to add *.h to Makefile...
10/25/91: Removed last traces of old getopt schizophrenia.
10/25/91: Changed *rotate to preserve modes of log files.
10/25/91: Added sessnowinit, scnowinit.
10/25/91: Added sessrotate, sclogrotate.
10/25/91: Fixed off-by-one on remotelen.
10/25/91: Changed wtmprotate to depend on config/wtmpfile.h properly.
[log is not complete before this point]
10/6/91: Added -xsS.
9/30/91: pty 4.0, second beta.


sigdfl.c
/* sigdfl.c, sigdfl.h: default signal library
Daniel J. Bernstein, brnstnd@nyu.edu.
No dependencies.
Requires BSD signal syscalls.
1/27/92: Added ABRT->IOT ifndef.
1/27/92: Switched from setting sv_flags to manually zeroing sigvecs.
7/18/91: Baseline. sigdfl 1.0, public domain.
No known patent problems.

Documentation in sigdfl.3.

If you port this to System V, you don't have to worry about stop
signals; you can just have sigdfl_tstp() and friends return -1 with
ENOTTY. You are, however, faced with an incomplete, nearly prehistoric
signal interface. Have fun.
*/

#include <signal.h>
#include "sigdfl.h"

static int cont = 0;

static sigcont() /* XXX: should declare with right signal type */
{
 cont = 1;
}

int sigdfl(sig)
int sig;
{
 int oldmask;
 struct sigvec oldvec;
 struct sigvec vec;
 struct sigvec contvec;
 char *x;

 if (sig == SIGCONT)
   return 0; /* strategy below simply cannot work for CONT */
 if ((sig == SIGURG) || (sig == SIGCHLD) || (sig == SIGIO)
#ifdef SIGWINCH
   || (sig == SIGWINCH)
#endif
    )
   return 0; /* the above signals are ignored */
 /* XXX: If we're still going now, and sig can be delivered without */
 /* killing the process and without stopping the process so that it'll */
 /* receive CONT later, then we will enter an infinite loop. [sigh] */
 /* XXX: put maximum time wastage on this? */
 oldmask = sigblock(0);
 sigblock(~0);
 /* now we won't receive any signals */
 x = (char *) &vec; while (x < sizeof(vec) + (char *) &vec) *x++ = 0;
 vec.sv_handler = SIG_DFL;
 vec.sv_mask = ~0;
 if (sigvec(sig,&vec,&oldvec) == -1)
   if ((sig != SIGSTOP) && (sig != SIGKILL))
     return -1;
 x = (char *) &vec; while (x < sizeof(vec) + (char *) &vec) *x++ = 0;
 vec.sv_handler = sigcont;
 vec.sv_mask = ~0;
 if (sigvec(SIGCONT,&vec,&contvec) == -1)
   return -1;
 cont = 0;
 if (kill(getpid(),sig) == -1)
   return -1;
 /* now a sig should be queued, and we have control over sig and CONT */
 /* exception: SIGSTOP and SIGKILL can't be blocked, so those signals
    might already have been delivered. in the SIGSTOP case, if we've
    reached this point, sigcont() might already have been run. that's
    why cont must be set to 0 before the kill(). */
 /* after this next bit we may receive sig and/or CONT */
 sigsetmask(~(sigmask(sig) | sigmask(SIGCONT)));
 /* in the near future, sig will in fact be received */
 while (!cont) /* dead loop until we receive CONT */
   ; /* XXX: there should be a syscall so we don't have to loop here */
 sigblock(~0);
 /* now we won't receive any signals */
 (void) sigvec(sig,&oldvec,&vec); /* we don't care if it fails */
 (void) sigvec(SIGCONT,&contvec,&vec);
 /* now signal handlers are back to normal */
 (void) sigsetmask(oldmask);
 return 0;
}

int sigdfl_tstp()
{
 return sigdfl(SIGTSTP);
}

int sigdfl_stop()
{
 return sigdfl(SIGSTOP);
}

int sigdfl_ttin()
{
 return sigdfl(SIGTTIN);
}

int sigdfl_ttou()
{
 return sigdfl(SIGTTOU);
}

int sigdfl_abrt()
{
#ifndef SIGABRT
#define SIGABRT SIGIOT
#endif
 return sigdfl(SIGABRT);
}
